pkg/chunkenc/gzip.go

```diff
@@ -477,73 +468,109 @@ func (li *listIterator) Close() error   { return nil }
 func (li *listIterator) Labels() string { return "" }
 
 type bufferedIterator struct {
-	s *bufio.Reader
+	s      *bufio.Reader
+	reader CompressionReader
+	pool   CompressionPool
 
-	curT   int64
-	curLog string
+	cur logproto.Entry
 
 	err error
 
-	buf    []byte // The buffer a single entry.
-	decBuf []byte // The buffer for decoding the lengths.
+	buf    *bytes.Buffer // The buffer a single entry.
+	decBuf []byte        // The buffer for decoding the lengths.
+
+	closed bool
+
+	filter logql.Filter
 }
 
-func newBufferedIterator(s *bufio.Reader) *bufferedIterator {
+func newBufferedIterator(pool CompressionPool, b []byte, filter logql.Filter) *bufferedIterator {
+	r := pool.GetReader(bytes.NewBuffer(b))
 	return &bufferedIterator{
-		s:      s,
-		buf:    make([]byte, 1024),
+		s:      BufReaderPool.Get(r),
+		reader: r,
+		pool:   pool,
+		filter: filter,
 		decBuf: make([]byte, binary.MaxVarintLen64),
 	}
 }
 
 func (si *bufferedIterator) Next() bool {
+	for {
+		ts, line, ok := si.moveNext()
+		if !ok {
+			si.Close()
+			return false
+		}
+		if si.filter != nil && !si.filter(line) {
+			continue
+		}
+		si.cur.Line = string(line)
+		si.cur.Timestamp = time.Unix(0, ts)
+		return true
+	}
+}
+
+// moveNext moves the buffer to the next entry
+func (si *bufferedIterator) moveNext() (int64, []byte, bool) {
 	ts, err := binary.ReadVarint(si.s)
 	if err != nil {
 		if err != io.EOF {
 			si.err = err
 		}
-		return false
+		return 0, nil, false
 	}
 
 	l, err := binary.ReadUvarint(si.s)
 	if err != nil {
 		if err != io.EOF {
 			si.err = err
-
-			return false
+			return 0, nil, false
 		}
 	}
 
-	for len(si.buf) < int(l) {
-		si.buf = append(si.buf, make([]byte, 1024)...)
+	if si.buf == nil {
+		si.buf = BytesBufferPool.Get()
 	}
 
-	n, err := si.s.Read(si.buf[:l])
+	for si.buf.Cap() < int(l) {
+		si.buf.Grow(int(l) - si.buf.Cap())
```

Comment:
See: https://github.com/prometheus/prometheus/blob/master/pkg/pool/pool.go Think it's better suited for this?

Reply:
Nice I'll look into it.

Reply:
The problem is log line size are unbounded. WDYT ?

Reply:
Yes, now that is the tradeoff that the promtheus-pool made. So it knows the length-buckets it wants to cache the lines of. Now if 99% of the lines are 5K, 1% are 100K, then we will be holding the 100K worth of memory for all blocks, while just allocating it for those 1% might make sense and keeping only 5K buffers stored, we might be having lower memory usage.

So I don't have answer here, but it might make sense to the use the pooling strategy from Prometheus.

Reply:
I'll give a try about this today.

Reply:
I see, so if we have one single 100k line it will create it but not add it to the pool. So it is also unbounded in way. Problem is I don't have a current benchmark that proves this is getting better. However I've seen that most line are below 4k but some are 22k. I'll try to improve one benchmark to catch the improvement.

Reply:
So I did the implementation and didn't notice any improvement however, by adding a test that send random size line I found something else very interesting.

We already talked about this but the line after was :
```
if n < int(l) {
		_, err := si.s.Read(si.buf.Bytes()[n:l])
		if err != nil {
			si.err = err
			return 0, nil, false
		}

	}
```

I realized for big line it will never read more than 4096 bytes at the time(because of the buffered reader below), thus even the second read was not enough which would cause the whole block to be totally off and it would send back `l` value not reasonable sometimes even crashing the whole benchmark.


I fixed it by doing this instead:

```
for n < int(l) {
		r, err := si.s.Read(si.buf.Bytes()[n:l])
		if err != nil {
			si.err = err
			return 0, nil, false
		}
		n += r
	}
```

The problem is gone. I'm going back to see if I can prove that with prometheus pool it is better now. I'm a bit scared of having put/get inside the moveNext though. Currently a single buffer is used for the whole block.

Reply:
I tried to implement it without clear improvement:

before
```
BenchmarkReadGZIP-16    	      10	 152987418 ns/op	158103939 B/op	   46406 allocs/op
````
after
````
BenchmarkReadGZIP-16    	      10	 192691303 ns/op	148492927 B/op	   59621 allocs/op
```

I would like to keep this for another PR, I don't really feel confident about that change.

Reply:
The buffer is doing a better job because we keep only one per block, while the promtheus pool need to be put then get the slice inside moveNext to get the correct size.